### PR TITLE
Feature/dsg 1510/dashboard ordering

### PR DESCRIPTION
**Pull request by jcsuarezm**
## Types of changes

Patch
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Documentation change (documentation only, will not require redeployment)
- [ ] Refactor or cleanup (functionality unchanged)
Minor
- [ ] New feature (non-breaking change which adds functionality)
Major
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have performed a self-review of my own code
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been merged and published in downstream modules

- [ ] I have confirmed all acceptance criteria defined by the story have been met

### CODE REVIEW

File: service/src/main/java/io/nuvalence/workmanager/service/service/TransactionDefinitionSetOrderService.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    private final TransactionDefinitionSetService transactionDefinitionSetService;
+
+    /**
+     * Updates the transaction definition set order.
+     *
+     * @param newOrder the new order
+     */
+    @Transactional
+    public void updateTransactionSetKeyOrder(List<String> newOrder) {
+        repository.findAll().stream()
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/workmanager/service/service/TransactionDefinitionSetOrderService.java:34]: would this be better to delete all elements in the table ?
My understanding you will replace the whole current list with the new list ?

File: service/src/main/java/io/nuvalence/workmanager/service/domain/transaction/TransactionDefinitionSet.java [hunk 2]
```diff
@@ -79,4 +79,11 @@ public class TransactionDefinitionSet implements UpdateTrackedEntity {
 
     @Column(name = "last_updated_timestamp")
     private OffsetDateTime lastUpdatedTimestamp;
+
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/workmanager/service/domain/transaction/TransactionDefinitionSet.java:82]: as you are going to query the ordering not through the transaction definition set,  is it necessary to have to have this eager loading entity here ?

File: service/src/main/java/io/nuvalence/workmanager/service/domain/transaction/TransactionDefinitionSet.java [hunk 3]
```diff
@@ -79,4 +79,11 @@ public class TransactionDefinitionSet implements UpdateTrackedEntity {
...
+    @Fetch(FetchMode.SELECT)
+    @OneToOne(
+            mappedBy = "transactionDefinitionSet",
+            cascade = CascadeType.ALL,
+            fetch = FetchType.EAGER)
+    private TransactionDefinitionSetOrder transactionDefinitionSetOrder;
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/workmanager/service/domain/transaction/TransactionDefinitionSet.java:88]: as you are going to query the ordering not through the transaction definition set, is it necessary to have to have this eager loading entity here ?

File: service/src/main/java/io/nuvalence/workmanager/service/service/DashboardConfigurationService.java [hunk 4]
```diff
@@ -17,9 +18,26 @@
...
 
     private final DashboardConfigurationRepository dashboardConfigurationRepository;
+    private final TransactionDefinitionSetOrderService transactionDefinitionSetOrderService;
 
+    /**
+     * Gets all dashboards that are expected to be listed.
+     * @return list of dashboards.
+     */
     public List<DashboardConfiguration> getAllDashboards() {
-        return (List<DashboardConfiguration>) dashboardConfigurationRepository.findAll();
```
Comment (deibys-nuvalence) [inline service/src/main/java/io/nuvalence/workmanager/service/service/DashboardConfigurationService.java:22]: not sure as it is not clear in requirements ...
the returned list should be driven based on the ordering of transaction definition sets ?

It is hard to interpret in the requirements if "Returns an ordered list of dashboard configurations" means only returned dashboard specified in the transaction set order list or the ordering refers to something else ?

**Review by JayStGelais — CHANGES_REQUESTED** [review]

File: service/src/main/java/io/nuvalence/workmanager/service/service/TransactionDefinitionSetOrderService.java (hunk 1 above)
Comment (jcsuarezm) [reply service/src/main/java/io/nuvalence/workmanager/service/service/TransactionDefinitionSetOrderService.java]: Thanks Jay and Deibys, a delete all has been introduced as well as some db constraint changes to allow a simpler code here.

File: service/src/main/java/io/nuvalence/workmanager/service/domain/transaction/TransactionDefinitionSet.java (hunk 3 above)
Comment (jcsuarezm) [reply service/src/main/java/io/nuvalence/workmanager/service/domain/transaction/TransactionDefinitionSet.java]: Thanks Jay and Deibys,this was modified to use @Formula to only load the sort order. This was chosen in place of implementing via join in order to prioritize code maintainability and simplicity as the performance impact was not considered significant. The later option would have required additional service and repository level code that would have added unnecessary complexity.

File: service/src/main/java/io/nuvalence/workmanager/service/service/DashboardConfigurationService.java (hunk 4 above)
Comment (jcsuarezm) [reply service/src/main/java/io/nuvalence/workmanager/service/service/DashboardConfigurationService.java:22]: Thank Deibys, this has been validated and is the expected behavior.

File: helm/work-manager/cerbos-policies/dashboard_configuration_resource.yaml [hunk 5]
```diff
@@ -3,7 +3,7 @@ resourcePolicy:
   version: "default"
   resource: dashboard_configuration
   rules:
-    - actions: ["view"]
+    - actions: ["view", "update"]
       roles:
-        - transaction-admin
+        - transaction-config-admin
```
[the remark this answers is not in the document]
Comment (jcsuarezm) [reply helm/work-manager/cerbos-policies/dashboard_configuration_resource.yaml:13]: Thanks Jay, the cerbos configuration has been made more granular to allow viewing for both roles, but update is only allowed for the config admin.

File: service/src/main/java/io/nuvalence/workmanager/service/service/DashboardConfigurationService.java [hunk 6]
```diff
@@ -17,9 +18,26 @@
...
+        List<DashboardConfiguration> listOfOrderedAndFilteredDashboards = new ArrayList<>();
+        transactionDefinitionSetOrderService.getTransactionDefinitionSetOrder().stream()
```
[the remark this answers is not in the document]
Comment (jcsuarezm) [reply service/src/main/java/io/nuvalence/workmanager/service/service/DashboardConfigurationService.java:29]: Thanks Jay, this has been simplified by using the proposed validation, reducing the calls to the db.

**Review by JayStGelais — APPROVED** [review]

**Pull request merged**